### PR TITLE
Fix #163

### LINKED ISSUE #163: Make Session loading (Deserialize) faster
[DistributedSession.cs](https://github.com/aspnet/Session/blob/02b975c59776722c0cc47542599a59bef9445440/src/Microsoft.AspNetCore.Session/DistributedSession.cs#L294) adds deserialized property values in a loop to a `Dictionary<EncodedKey, byte[]> _store`.

This causes continuous dictionary resizing (which uses prime-number capacities). However, this resizing is wasteful, since `Deserialize()` already knows the required capacity (`expectedEntries`).

`_store` should be initialised with a known capacity inside `Deserialize()`, instead if being initialised with an unknown capacity inside ctor.

### PR DESCRIPTION
**Pull request by hishamco**

### CODE REVIEW

**Comment by Tratcher** [comment]
This is the same as what the last guy tried and it caused lots of null refs.
https://github.com/aspnet/Session/pull/196/files

The failure code path likely needs to initialize the collection as well.

**Comment by hishamco** [comment]
Yep, that's what I did right now on my laptop, I will update this PR now ..

File: src/Microsoft.AspNetCore.Session/DistributedSession.cs [hunk 1]
```diff
@@ -70,8 +70,8 @@ public class DistributedSession : ISession
             _idleTimeout = idleTimeout;
             _ioTimeout = ioTimeout;
             _tryEstablishSession = tryEstablishSession;
-            _store = new Dictionary<EncodedKey, byte[]>();
             _logger = loggerFactory.CreateLogger<DistributedSession>();
+            _store = new Dictionary<EncodedKey, byte[]>();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Session/DistributedSession.cs]: What if you moved this here instead:
https://github.com/hishamco/Session/blob/06b2054f1761803c7de7b8b32370df36cdd32471/src/Microsoft.AspNetCore.Session/DistributedSession.cs#L181

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Session/DistributedSession.cs]: Yes, in that else block.

Comment (hishamco) [reply src/Microsoft.AspNetCore.Session/DistributedSession.cs]: This is a good option too

File: src/Microsoft.AspNetCore.Session/DistributedSession.cs [hunk 2]
```diff
@@ -178,6 +177,7 @@ private void Load()
                     }
                     else if (!_isNewSessionKey)
                     {
+                        _store = new Dictionary<EncodedKey, byte[]>();
```
Comment (muratg) [inline src/Microsoft.AspNetCore.Session/DistributedSession.cs]: This looks good, but will it break anything else I wonder.

Comment (hishamco) [reply src/Microsoft.AspNetCore.Session/DistributedSession.cs]: Yep, it breaks 😞

**Comment by Tratcher** [comment]
Closing until someone comes up with the right fix.

**Pull request closed without merging**